Missing components type with `vue-facing-decorator`

### Vue - Official extension or vue-tsc version

2.1.2

### VSCode version

I use WebStorm

### Vue version

3.4.38

### TypeScript version

5.5.4

### System Info

```shell
System:
    OS: macOS 14.6.1
    CPU: (8) arm64 Apple M1
    Memory: 80.02 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.15.1 - /usr/local/bin/node
    npm: 10.7.0 - /usr/local/bin/npm
    pnpm: 7.13.5 - /usr/local/bin/pnpm
  Browsers:
    Chrome: 128.0.6613.85
    Safari: 17.6
```


### Steps to reproduce

Launch command:  
`vue-tsc --noEmit --skipLibCheck -p tsconfig.json --composite false`


### What is expected?

It's expected the build is correctly completed

### What is actually happening?

Every reference of custom components in other custom component templates are marked as errors this way:

```
src/components/partials/ZBaseDatePicker.vue:55:10 - error TS2339: Property 'ZPlainButton' does not exist on type 'Required<{} & __VLS_WithComponent<"ZPlainButton", { "i18n-t": new () => { $props: VNodeProps & TranslationProps; }; "i18n-d": new () => { $props: VNodeProps & DatetimeFormatProps & BaseFormatProps; }; ... 8 more ...; Teleport: { ...; }; } & this, "ZPlainButton", "zPlainButton", "z-plain-button"> & __VLS_WithComponen...'.

55         <z-plain-button

```

### Link to minimal reproduction

_No response_

### Any additional comments?

_No response_

--- a/packages/language-core/lib/codegen/script/index.ts
+++ b/packages/language-core/lib/codegen/script/index.ts
@@ -114,8 +114,8 @@ export function* generateScript(options: ScriptCodegenOptions): Generator<Code, 
 			}
 			else {
 				yield generateSfcBlockSection(options.sfc.script, 0, classBlockEnd, codeFeatures.all);
-				yield `__VLS_template = () => {`;
-				const templateCodegenCtx = yield* generateTemplate(options, ctx, true);
+				yield `__VLS_template = () => {${newLine}`;
+				const templateCodegenCtx = yield* generateTemplate(options, ctx);
 				yield* generateComponentSelf(options, ctx, templateCodegenCtx);
 				yield `},${newLine}`;
 				yield generateSfcBlockSection(options.sfc.script, classBlockEnd, options.sfc.script.content.length, codeFeatures.all);
@@ -140,7 +140,7 @@ export function* generateScript(options: ScriptCodegenOptions): Generator<Code, 
 
 	if (!ctx.generatedTemplate) {
 		yield `function __VLS_template() {${newLine}`;
-		const templateCodegenCtx = yield* generateTemplate(options, ctx, false);
+		const templateCodegenCtx = yield* generateTemplate(options, ctx);
 		yield `}${endOfLine}`;
 		yield* generateComponentSelf(options, ctx, templateCodegenCtx);
 	}

--- a/packages/language-core/lib/codegen/script/scriptSetup.ts
+++ b/packages/language-core/lib/codegen/script/scriptSetup.ts
@@ -288,7 +288,7 @@ function* generateSetupFunction(
 	yield* generateComponentProps(options, ctx, scriptSetup, scriptSetupRanges, definePropMirrors);
 	yield* generateModelEmits(options, scriptSetup, scriptSetupRanges);
 	yield `function __VLS_template() {${newLine}`;
-	const templateCodegenCtx = yield* generateTemplate(options, ctx, false);
+	const templateCodegenCtx = yield* generateTemplate(options, ctx);
 	yield `}${endOfLine}`;
 	yield* generateComponentSelf(options, ctx, templateCodegenCtx);
 	yield `type __VLS_TemplateResult = ReturnType<typeof __VLS_template>${endOfLine}`;

--- a/packages/language-core/lib/codegen/script/template.ts
+++ b/packages/language-core/lib/codegen/script/template.ts
@@ -8,18 +8,11 @@ import { generateStyleScopedClasses } from '../template/styleScopedClasses';
 import type { ScriptCodegenContext } from './context';
 import { codeFeatures, type ScriptCodegenOptions } from './index';
 
-function* generateTemplateCtx(
-	options: ScriptCodegenOptions,
-	isClassComponent: boolean
-): Generator<Code> {
+function* generateTemplateCtx(options: ScriptCodegenOptions): Generator<Code> {
 	const exps = [];
 
-	if (isClassComponent) {
-		exps.push(`this`);
-	}
-	else {
-		exps.push(`{} as InstanceType<__VLS_PickNotAny<typeof __VLS_self, new () => {}>>`);
-	}
+	exps.push(`{} as InstanceType<__VLS_PickNotAny<typeof __VLS_self, new () => {}>>`);
+
 	if (options.vueCompilerOptions.petiteVueExtensions.some(ext => options.fileBaseName.endsWith(ext))) {
 		exps.push(`globalThis`);
 	}
@@ -116,8 +109,7 @@ export function* generateTemplateDirectives(options: ScriptCodegenOptions): Gene
 
 export function* generateTemplate(
 	options: ScriptCodegenOptions,
-	ctx: ScriptCodegenContext,
-	isClassComponent: boolean
+	ctx: ScriptCodegenContext
 ): Generator<Code, TemplateCodegenContext> {
 	ctx.generatedTemplate = true;
 
@@ -125,7 +117,7 @@ export function* generateTemplate(
 		scriptSetupBindingNames: new Set(),
 		edited: options.edited,
 	});
-	yield* generateTemplateCtx(options, isClassComponent);
+	yield* generateTemplateCtx(options);
 	yield* generateTemplateComponents(options);
 	yield* generateTemplateDirectives(options);
 	yield* generateTemplateBody(options, templateCodegenCtx);

--- a/test-workspace/tsc/passedFixtures/vue3/#4841/main.vue
+++ b/test-workspace/tsc/passedFixtures/vue3/#4841/main.vue
@@ -1,0 +1,7 @@
+<script lang="ts">
+export default class {}
+</script>
+
+<template>
+	<Comp></Comp>
+</template>

